PPing core improvements

simosund opened the pull request:
Adds RTT-based sampling, ICMP echo support and making ePPing work in both directions, i.e. doing both timestamping and matching on both ingress and egress.

Also contains a couple of commits with fixes to the BPF code.

Commits based [PR#23](https://github.com/xdp-project/bpf-examples/pull/23) but slightly cleaned up.

This PR is based on top of [PR#31](https://github.com/xdp-project/bpf-examples/pull/31). On request I could try and rebase it on top of master instead (will likely cause some conflicts), but I expect PR#31 to be merged first at which point I would have to rebase this to the new master which would effectively undo all the work in the first rebase.

tohojo commented:
Right, so I'm afraid I won't have time to properly review this before the holidays, sorry about that :(

Will circle back to it in January :)

simosund commented:
No worries, I've gone on vacation myself now. Merry christmas and happy new year!

tohojo commented:
Hmm, seems Github didn't pick up on the first bit of this being merged. Could you please do a manual rebase onto master and re-push this?

simosund commented:
> Hmm, seems Github didn't pick up on the first bit of this being merged. Could you please do a manual rebase onto master and re-push this?

Hmm strange, maybe because this series of commits was missing the merge-commit? Anyways, have rebased on the new merged master, so should hopefully be fine now

tohojo reviewed (changes requested):
Most of the comments are relatively minor, I guess. My main overall comment is that you still have a tendency to lump unrelated things together in single commits, which makes reviewing harder than it needs to be. When splitting things into commits, try to think of it this way: each commit should be self-contained in that it doesn't require any context other than the code changes themselves and its commit message. You should generally be making the commits as small as possible, but not so small that they lack context. It's fine to do partial refactorings, though, as long as you call out upcoming changes ("change X in preparation for later doing Y"). In general, think of it as minimising the amount of stuff the reviewer has to keep in their head at once (which is also helpful when writing commit messages; they are there to help the reviewer, and yourself when you later go back and run `git blame` to figure out why some code is the way it is).

I'm also not a huge fan of the "issues with this commit" items at the bottom of each commit message. I view them as a bit of a cop-out: the issues were obviously not important enough to fix, or you would have done that already. And the commit message should be describing the change you are making, not all the other stuff you thought about while making that change :) Tracking TODO items is better done in a TODO file, or an issue tracker (and feel free to use the issue tracker in this repository, BTW). If you do use a TODO file, add a separate commit at the end of your series updating that; this also helps you scope out the series (or pull request, as it were). Like "This series did X, Y and Z, and now I'm updating the TODO list with future items A, B and C that I will tackle in a subsequent series).

Oh, and the above comments are mostly for future reference - you don't need to restructure the commits in this series! And yeah, splitting changes up in readable chunks can be a lot of tedious work, but it gets easier with practice; you'll learn to think in terms of the changes and how you will describe them, which can also improve the code (since you'll sometimes think about and fix issues while you're doing such a split).

simosund commented:
Thanks for the extensive feedback, will try to fix most of it tomorrow.

As for "issues with this commit" part, I don't intend for them to be cop outs, but rather I want to highlight things that I think might be potential issues. Genereally I don't consider them critical (in that case I would hopefully have fixed them), but by making them very visibel my hope is essentially that you might disagree and find that it completely breaks the application, in which case we can have a discussion about how to fix it ASAP. I also want to note them down while I still remeber them.

That said, I realize that the commit messages might not be the best place for that. While I actually have a TODO list I've been pretty bad at updating it, and generally view it more as a list of additional features rather than for tracking potential issues. I also have a list of potential concurrency issues in the README (which was initially spread out as a bunch of comments in the code), so maybe it would make sense to consolidate all that in a dedicated document, or add a separate section in the TODO list with them? While the issue tracker might also be a good way to do that, I have a feeling that it should mainly be used for issues in the current master, not for potential issues within a PR which may have been fixed by the time they're merged?

simosund commented on pping/pping_kern.c:
```diff
@@ -371,20 +372,25 @@ static void pping_egress(void *ctx, struct parsing_context *pctx)
+		}
 
+		f_state = bpf_map_lookup_elem(&flow_state, &p_id.flow);
 		if (!f_state) // Creation failed
```
Hmm, ocassionally emit a warning to stderr about failing to track flow due to flow-state map being full? Definitly need to make sure that it goes out with a limited frequency though, so that not every single packet triggers a warning message which would tank performance in a period where PPing is likely already under a heavy load.

Ideally I guess one would perhaps get one message per flow which failed to create a new state, but I don't really see a way to track that I've already emitted a warning for a particular flow without keeping some state for that flow...but the reason for emitting this warning to begin with is that we're out of space for tracking flows. 

So simply some time-based limit? Although I'm not really sure what frequency would make sense here. 1000 warnings/sec? 100 warnings/sec? 10 warnings/sec?

tohojo replied:
Once a second would probably be fine, at least in interactive mode... You could just keep a timestamp in a global variable and have logic like:

```c
if (!f_state) { // creation failed
  if (last_warn_time < now - WARN_INTERVAL) {
    emit_warning(); // send a perf event of a special type, for instance
    last_warn_time = now;
  }
}
```

simosund replied:
Have added a new commit at the end that should address this. In addition to sending a warning if it fails to create a flow state entry (due to the flow state map being full), it also sends a warning in case it fails to create a timestamp entry.

Both are capped to only emitting one warning per second (currently hard-coded, but could of course be made configurable in case you believe it's something the user might want to change), but tracked separately for the flow state and timestamp map.

simosund commented on pping/pping.c:
```diff
@@ -106,6 +103,7 @@ static const struct option long_options[] = {
 	{ "cleanup-interval", required_argument, NULL, 'c' }, // Map cleaning interval in s, 0 to disable
 	{ "format",           required_argument, NULL, 'F' }, // Which format to output in (standard/json/ppviz)
 	{ "ingress-hook",     required_argument, NULL, 'I' }, // Use tc or XDP as ingress hook
+	{ "localfilt-off",    no_argument,       NULL, 'l' }, // Disable local filtering (will start to report "internal" RTTs)
```
Agreed. Think I first called it just localfilt, but figured that was a very misleadning name as the option turns it off not on, so my quick-fix was to just append off to the end =) Wanted to keep -l to be consistent with Kathie's PPing, but guess that's still fairly reasonable with `--local-filt` (which is also a much more descriptive name).

simosund replied:
See I ended up writing `--local-filt` in my previous comment, but of course meant to change it to your suggested `--include-local` which I've now also done.

simosund commented on pping/pping_kern.c:
```diff
@@ -25,6 +25,8 @@
 #define AF_INET6 10
 #define MAX_TCP_OPTIONS 10
 
+#define IPV6_FLOWINFO_MASK __cpu_to_be32(0x0FFFFFFF)
```
Don't really know. Pretty much just copied how you filled in the `bpf_fib_lookup` in [sample/bpf/xdp_fwd_kern.c](https://github.com/torvalds/linux/blob/26291c54e111ff6ba87a164d85d4a4e134b7315c/samples/bpf/xdp_fwd_kern.c#L43-L100) and then you used this thing so I pretty much followed blindly. Should probably take some time figuring out how `bpf_fib_lookup` struct is actually used like, but remeber that I thought the struct seemed rather complicated and was unsure which fields needed to be set and with what.

tohojo replied:
Yeah, good idea ;)
It's commented in the struct, to a certain extent: https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/bpf.h#L6121

simosund replied:
Have added a comment that this is for getting the flow label + traffic class from the IPv6 header which apparently is used in the lookup.

simosund commented on pping/pping_kern.c:
```diff
@@ -351,86 +392,154 @@ static bool is_rate_limited(__u64 now, __u64 last_ts, __u64 rtt)
+	}
+
+	lookup.family = p_info->pid.flow.ipv;
+	lookup.l4_protocol = p_info->pid.flow.proto;
```
No idea, as above, I pretty much just copy-pasted the one example I found of how this thing was used, and there the l4 protocol is filled in while the ports are set to 0.

simosund replied:
Still not sure about this one honestly, both examples of it that I've found ([xdp_fwd_kern](https://github.com/torvalds/linux/blob/c36c04c2e132fc39f6b658bf607aed4425427fd7/samples/bpf/xdp_fwd_kern.c#L71-L97) and [test_tc_neigh_fib](https://github.com/torvalds/linux/blob/c36c04c2e132fc39f6b658bf607aed4425427fd7/tools/testing/selftests/bpf/progs/test_tc_neigh_fib.c#L24-L76)) does it like that. Guess there might be forwarding rules based solely on the protocol, i.e. forward TCP traffic but not UDP traffic, without knowledge of the ports. So have left it like this for now.

Question is if it really matters for our use-case though, which is to find out if the packet is going to the same machine that PPing is running on or not. Perhaps it makes more sense to only provide it with IP addresses, as I guess the protocol shouldn't really affect if the local machine is the intended destination of the packet or not.

simosund commented on pping/pping_kern.c:
```diff
@@ -517,10 +648,11 @@ int pping_tc_egress(struct __sk_buff *skb)
 		.data_end = (void *)(long)skb->data_end,
 		.pkt_len = skb->len,
 		.nh = { .pos = pctx.data },
+		.ifindex = skb->ingress_ifindex,
```
Good point.

simosund commented on pping/pping_kern.c:
```diff
@@ -225,15 +285,21 @@ static int parse_packet_identifier(struct parsing_context *ctx,
+	if (p_id->flow.proto == IPPROTO_TCP)
+		err = parse_tcp_identifier(ctx, &saddr->port, &daddr->port, fei,
+					   &p_id->identifier);
+	else if (p_id->flow.proto == IPPROTO_ICMPV6 &&
```
Hmm, while having it optional definitly makes sense, I would assume the "all" would be the default, like it would be for ex pcap (i.e. you can add filters for different protocols if you want, but by default it captures everything it can). 

But are you suggesting we add a flag for each supported protocol the user wants to track (plus a flag for all), and then assume TCP if user specifies no flags? Or should TCP always be included by default and we instead have a flag if you explicitly want to exclude TCP, whereas for all the other protocols the flag is used to include it?

abrunstrom replied:
I think the first makes most sense. A flag for each protocol, a flag for all, and if nothing is specified for the protocol part use TCP as default. But if one or several protocols are explicitly specified this overrides the default and then you use what is specified. The default should be what you expect to be most commonly used I think.

tohojo replied:
I don't have any strong opinions on what the default should be, just that it should be configurable :)

simosund replied:
Have gone ahead and added separate flags for enabling TCP (`--tcp`/`-T`), ICMP (`--icmp`/`-C`) and a shorthand for all supported protocols (`--all`/`-A`). If no flag is specified the user is notified that it assumes (only) TCP.

tohojo commented:
Hmm, if the issues you point out are meant for review, I think it's better to include them in the pull request description (which would be the cover letter in an email submission). If they're in the commit, they're a bit ambiguous as well ("issues with this commit" - does that mean they will be fixed with a subsequent commit?). And if they're longer-term things, put them in the TODO or somewhere similar.

And yes, if you already have fixed queued up in a subsequent pull request, there's no reason to open an issue. But then that is what the text in the PR description should say. If they're longer term issues, they can be tracked in issues just fine - and if you're not sure, put them in the PR description and ask :)

simosund commented on pping/pping_kern.c:
```diff
@@ -371,20 +372,25 @@ static void pping_egress(void *ctx, struct parsing_context *pctx)
+					BPF_NOEXIST) == 0) {
+			new_flow = true;
+
+			if (fe.event_info.event != FLOW_EVENT_OPENING) {
```
So after cursing my old self any trying to figure out what I was thinking at the time, I believe that I initially did this in some attempt to optimize the code by only filling in the fields that where needed when they were needed. With the commit that made PPing "symmetrical" (doing both timestamping and matching on both ingress and egress) I essentially gave up on this idea, but for some reason kept passing this struct around between functions (even if the function itself filled all the members, meaning it could just as well allocate it itself).

I have addressed this in the commit that makes PPing symmetrical, so it is now only allocated and filled inside functions which directly push it to the perf-buffer (as the only reason I ever fill it is so I can push it on the perf-buffer). I've opted to not go back and try and fix it before this commit however, as I rather not spend energy rewriting code that is overwritten by later changes in this PR anyways (and it was also a problem which existed since before this PR).

simosund commented on pping/pping.h:
```diff
@@ -135,4 +138,16 @@ union pping_event {
 };
 
+/*
+ * Copies the src to dest, but swapping place on saddr and daddr
```
Have changed the comment to instead try to motivate what the function is useful for.

simosund commented on pping/pping_kern.c:
```diff
@@ -337,6 +375,9 @@ static __u64 calculate_srtt(__u64 prev_srtt, __u64 rtt)
 }
 
+/*
+ * Return true if flow should wait with timestamping due to rate limit
```
Have removed this comment.

simosund commented:
Hi again @tohojo. Think I've addressed most of your comments now (other than the main one about grouping unrelated changes together in the commits, as I've not split up any of the existing commits...). I have however removed the "issues with this commit" parts from the commit messages.

Most of the commits before "timestamping + matching on both ingress and egress" are largely the same, just that I've added an additional commit there which adds command line flags for which protocols PPing should track.

Have done some additional refactoring in the ingress+egress commit to make a separate function for sending flow-events (which allocate their own flow-event structs rather than passing the same one around) and some other minor fixes and cleanup. The "consider flow opened on reply" commit has also undergone a few changes to be more consistent with the changes made in ingress+egress commit.

Finally, I've added a couple of commits which adds warnings if case PPing fails to create map entries, as well as a separate commit where I add section on "potential problems" to the TODO document. There I've summarized the various issues were previously part of the commit messages.

Finally, some of the other notes that were part of the commit messages before but would probably fit better in the PR description are:

- PPing is probably less performant now as it essentially does twice the amount of work (minus parsing the packet), with both timestamping and matching on each packet as well as updating the flow state in both directions.
- The verifier has a much easier time verifying the BPF programs now (processes much fewer instructions), despite them being more complex.
- PPing no longer works on a 5.4 kernel (or at least it didn't at the time of the original commits).

tohojo commented on pping/pping.c:
```diff
@@ -985,6 +1000,16 @@ int main(int argc, char *argv[])
+	if (config.bpf_config.track_icmp &&
+	    config.output_format == PPING_OUTPUT_PPVIZ)
+		fprintf(stderr,
+			"Warning: ppviz format mainly intended for TCP traffic, but may now include ICMP traffic as well\n");
```
Will it work at all, though? If not, probably better to just error out and refuse to run instead of just printing a warning...

tohojo replied:
Right, if you *can* meaningfully visualise it with ppviz, that's fine; but if ppviz would choke on the output anyway, running in that mode wouldn't really make much sense...

simosund replied:
Well, ppviz would visualize it the same way as it would a TCP stream (as far as ppviz is concerned everything is a TCP). How meaningful you find that visualization I guess can be debated, in general I find ppviz a bit clunky to use, but I don't think it's really any more or less meaningful for ICMP than it is for TCP.

![flent-example](https://user-images.githubusercontent.com/72788896/153391265-b2d70804-872b-4ce3-9608-57437325a2a5.png)

In the above example the first/bottom flow (10.11.1.2:7) is a normal ping instance, the middle flow (10.11.1.2:8) is a ping flood and the last/top flow (10.11.1.2:48762) is an iperf3 TCP flow (the 10.11.1.2:48760 flow is the "control" flow from iperf3 before the actual data stream starts).

I'm not sure I follow why ppviz would choke on ICMP output. A TCP stream will generate around 1000 RTT reports per second, while your typical ping only sends a single IMCP echo per second. By default ppviz will aggregate data in 10s windows, so even with just 1 report/s I still think you get somewhat meaningful reports. I guess a ping flood could maybe generate a large amount of reports, but I haven't really done any performance testing of ppviz to determine how much data it can handle, and by default ePPing will only generate 10 reports/s per flow anyways.

Overall I feel like it's unecessary to explicitly prevent users from running the ppviz format with anything other than TCP traffic because there may exist scenarios where it's not optimal. If we don't allow it because ppviz maybe will struggle with the amount of output from ping floods (although hopefully they're not that common on most networks), then should we also terminate the application if it's in ppviz mode when there's a large amount of councurrent TCP flows which would in turn also generate a lot of reports?

tohojo replied:
Totally fine, let's keep it :)

tohojo commented on pping/pping.c:
```diff
@@ -985,6 +1000,16 @@ int main(int argc, char *argv[])
 
+	if (!config.bpf_config.track_tcp && !config.bpf_config.track_icmp) {
+		config.bpf_config.track_tcp = true;
+		fprintf(stderr, "Warning: No protocol specified, assuming TCP\n");
```
So with this the common case will emit a warning, which seems a bit annoying. Maybe it's better to turn it into an informative message, like "starting pping in $MODE mode on $IFACE" or something to that effect?

simosund replied:
Yhea, emitting a warning for the default case is probably not optimal. However, I would still prefer to emit that message on stderr rather than stdout, as for stdout it may interfere with the JSON parsing (if using JSON output). I guess an alternative could be to simply add a very simple JSON entry instead, like ex. `{"info": "starting pping in mode..."}` but as the JSON output is not pretty-printed it will probably be a bit hard to read (and may take a while to display on terminal due to line buffering). I guess another approach is to simply skip that message if using JSON mode.

tohojo replied:
It's fine to output the message to stderr even if it's not strictly an "error". So keep doing that, but change it to a less-scary status message :)

tohojo commented on pping/pping.c:
```diff
@@ -235,6 +240,16 @@ static int parse_arguments(int argc, char *argv[], struct pping_config *config)
+		case 'C':
+			config->bpf_config.track_icmp = true;
+			break;
+		case 'A':
```
Since -T and -C are not exclusive, the -A mode is not strictly needed (the user can just supply both modes). I'd say drop it until we have enough modes that this becomes tedious :)

simosund replied:
Sure, guess writing `-TC` is not really much more work than writing `-A`. Although I'm not sure I understand the reason why not having the `--all` option right now and (potentially) adding it back later is better?

As a side note, think next protocol to be added should have the flag `-P` so you can write `-TCP` (to missleadingly get both TCP, ICMP and something else) =).

tohojo replied:
Well, you lock yourself into using the character for that purpose, and it's more options for the user to read through when looking at the help output... So not a huge deal, but if there's no reason to have it, you might as well save it for when it's actually useful? :)

simosund replied:
Sure, seems reasonable.

tohojo reviewed (changes requested):
A few nits on the mode selection, but otherwise I think we should get this merged so you can make progress on other items :)
